platform/admin/core/admin-core-appservice/src/main/java/org/codice/ddf/admin/application/service/migratable/TaskList.java

```diff
@@ -107,16 +115,62 @@ public void add(Operation op, String id, Predicate<ProfileMigrationReport> task)
   }
 
   /**
-   * Increases the number of attempts left for the specified operation
+   * Adds a compound task to be executed for a specific operation. This method expects to be called
+   * consistently for each subtasks of a given operation as the container and the compound task will
+   * only be created and registered the first time this method is called. The created container will
+   * be supplied to the provided accumulator each time the method is called to provide a way for the
+   * client of this API to accumulate information for each subtasks that will be later provided to
+   * the registered compound task when it is executed.
+   *
+   * <p><i>Note:</i> It is recommended to call this method for a given operation in one single place
+   * in your code to ensure consistency for the container type and the compound tasks for each
+   * subtasks being added.
+   *
+   * <p>
    *
-   * @param op the operation to increase the number of attempts for
-   * @throws IllegalArgumentException if <code>op</code> is <code>null</code>
+   * <pre><code>
+   *   private final TaskList tasks = new TaskList();
+   *   ...
+   *   tasks.add(
+   *     Operation.INSTALL,
+   *     name,
+   *     HashSet<MyObjectClass>::new,
+   *     (objects, r) -> [do something with the accumulated objects],
+   *     (n, objects) -> objects.add(obj));
+   * </code></pre>
+   *
+   * @param <T> the type of the container used to accumulate recorded subtasks
+   * @param op the operation for which to record the subtask
+   * @param id a unique identifier for the subtask (e.g. feature name)
+   * @param containerFactory a supplier which returns a new, empty container object (only called the
+   *     first time this method is called)
+   * @param task the compound task to add for later execution (the task will only be registered the
+   *     first time this method is called for a given operation)
+   * @param accumulator a consumer capable of accumulating data related to this particular subtask
+   *     in the provided container for later execution by the compound task; it will receive the
+   *     subtask id an the container
+   * @throws IllegalArgumentException if <code>op</code>, <code>id</code>, </code><code>
+   *     containerFactory</code>, <code>task</code>, <code>accumulator</code> is <code>null</code>
    */
-  public void increaseAttemptsFor(Operation op) {
+  public <T> void add(
```

Comment:
I am open to suggestions. 

This class came from multiple passes at the original design which was way more complicated. It simplified a lot the whole thing. The complexity comes from the fact that to work well with all the inter-dependency issues that surfaces with Karaf and bundle/feature operation, we are forced to do multiple passes. Without this class to more easily categorize tasks to execute on a per operation basis, I would be forced to keep track of that logic elsewhere. That was not simpler at all; in fact it was more complicated to understand.

By designing this class as a completely independent utility, it makes the logic of all the other classes much simpler so it is definitely a case of give and take. I opted for such a utility classes which can be more easily understood. It is meant to track tasks to execute later. It categorizes them per operation and only executes the highest registered operation at a time. In this PR, the notion of accumulating multiple subtasks together to execute one single task (i.e. a compound task) at the time we execute was required to deal with feature installation and uninstallation to take advantage of Karaf's dependency resolution which otherwise would require us to compute all of that (installing one feature at a time was not stable and would fail 50% of the time with the exact same conditions). In another PR I am working on, I am able to reuse this functionality to wait for all bundles affected by executing tasks to stabilize.

Let's not forget that if Karaf wasn't so complicated to use or play with, a lot of this complexity would not be required. Let's just look at how we are force to deal with apps today in the ApplicationServiceImpl or the config repo to see that any time we deal with Karaf, we are forced to add complexity on our side.